src/librustc/hir/mod.rs

```diff
@@ -444,71 +446,54 @@ pub enum TraitBoundModifier {
 pub type TyParamBounds = HirVec<TyParamBound>;
 
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
-pub struct TyParam {
-    pub name: Name,
+pub enum GenericParamKind {
+    /// A lifetime definition, eg `'a: 'b + 'c + 'd`.
+    Lifetime {
+        /// Either "'a", referring to a named lifetime definition,
+        /// or "" (aka keywords::Invalid), for elision placeholders.
+        ///
+        /// HIR lowering inserts these placeholders in type paths that
+        /// refer to type definitions needing lifetime parameters,
+        /// `&T` and `&mut T`, and trait objects without `... + 'a`.
+        name: LifetimeName,
```

Comment:
I'd be nice to share the `name` field, like in `ty::GenericParamDef`, I think the only prerequisite is to get rid of `Fresh` and use `'0`, `'1`, etc. as names, and/or try to figure out how to best deduplicate with synthetic type parameters.